dkist/asdf_maker/helpers.py

```diff
@@ -0,0 +1,182 @@
+import os
+
+import asdf
+import numpy as np
+import astropy.units as u
+from astropy.time import Time
+from astropy.io import fits
+from astropy.io.fits.hdu.base import BITPIX2DTYPE
+from astropy.modeling.models import (Shift, Multiply, AffineTransformation2D, Pix2Sky_TAN,
+                                     RotateNative2Celestial, Linear1D)
+from gwcs.lookup_table import LookupTable
+from asdf.tags.core.external_reference import ExternalArrayReference
+
+
+__all__ = ['make_asdf', 'time_model_from_date_obs', 'linear_time_model', 'linear_spectral_model',
+           'spatial_model_from_quantity', 'spatial_model_from_header', 'references_from_filenames']
+
+
+def references_from_filenames(filenames, array_shape, hdu_index=0, relative_to=None):
+    """
+    Given an array of paths to FITS files create a set of nested lists of
+    `asdf.external_reference.ExternalArrayReference` objects with the same
+    shape.
+
+    Parameters
+    ----------
+
+    filenames : `numpy.ndarray`
+        An array of filenames, in numpy order for the output array (i.e. ``.flat``)
+
+    array_shape : `tuple`
+        The desired output shape of the reference array. (i.e the shape of the
+        data minus the HDU dimensions.)
+
+    hdu_index : `int` (optional, default 0)
+        The index of the HDU to reference. (Zero indexed)
+
+    relative_to : `str` (optional)
+        If set convert the filenames to be relative to this path.
+    """
+
+    filenames = np.asanyarray(filenames)
+    reference_array = np.empty(array_shape, dtype=object)
+    if filenames.size != reference_array.size:
+        raise ValueError(f"An incorrect number of filenames ({filenames.size})"
+                         f" supplied for array_shape ({array_shape})")
+
+    for i, filepath in enumerate(filenames.flat):
+        with fits.open(filepath) as hdul:
+            hdu = hdul[hdu_index]
+            dtype = BITPIX2DTYPE[hdu.header['BITPIX']]
+            # hdu.shape is already in Python order
+            shape = tuple(hdu.shape)
+
+            # Convert paths to relative paths
+            relative_path = filepath
+            if relative_to:
+                relative_path = os.path.relpath(filepath, relative_to)
+
+            reference_array.flat[i] = ExternalArrayReference(
+                relative_path, hdu_index, dtype, shape)
+
+    return reference_array.tolist()
+
+
+def spatial_model_from_header(header):
+    """
+    Given a FITS compliant header with CTYPEx,y as HPLN, HPLT return a
+    `~astropy.modeling.CompositeModel` for the transform.
+
+    This function finds the HPLN and HPLT keys in the header and returns a
+    model in Lon, Lat order.
+    """
+    latind = None
+    lonind = None
+    for k, v in header.items():
+        if isinstance(v, str) and "HPLN" in v:
```

Comment:
this should check for `-TAN` or else it should be smart about choosing the projection model below.

Reply:
done